broadcast/src/linked/signer/actor.rs

```diff
@@ -0,0 +1,939 @@
+    mailbox_receiver: mpsc::Receiver<Message>,
+
+    ////////////////////////////////////////
+    // Storage
+    ////////////////////////////////////////
+    journal_entries_per_section: u64,
+    journal_replay_concurrency: usize,
+    journal_naming_fn: fn(&PublicKey) -> String,
+    journals: HashMap<PublicKey, Journal<B, E>>,
+
+    ////////////////////////////////////////
+    // State
+    ////////////////////////////////////////
+    tip_man: TipManager,
+
+    // Handles acknowledgements for chunks.
+    ack_man: AckManager,
+
+    // The current epoch.
+    epoch: Epoch,
+}
+
+impl<
+        B: Blob,
+        E: Clock + Spawner + Storage<B>,
+        C: Scheme,
+        H: Hasher,
+        A: Application<Context = Context>,
+        Z: Collector<Context = Context>,
+        S: ThresholdCoordinator<Index = Epoch, Share = group::Share, Identity = poly::Public>,
+    > Actor<B, E, C, H, A, Z, S>
+{
+    pub fn new(runtime: E, cfg: Config<C, H, A, Z, S>) -> (Self, Mailbox) {
+        let (mailbox_sender, mailbox_receiver) = mpsc::channel(cfg.mailbox_size);
+        let mailbox = Mailbox::new(mailbox_sender);
+        let result = Self {
+            runtime,
+            crypto: cfg.crypto,
+            hasher: cfg.hasher,
+            coordinator: cfg.coordinator,
+            application: cfg.application,
+            collector: cfg.collector,
+            chunk_namespace: encoder::chunk_namespace(&cfg.namespace),
+            ack_namespace: encoder::ack_namespace(&cfg.namespace),
+            refresh_epoch_timeout: cfg.refresh_epoch_timeout,
+            refresh_epoch_deadline: None,
+            rebroadcast_timeout: cfg.rebroadcast_timeout,
+            rebroadcast_deadline: None,
+            epoch_bounds: cfg.epoch_bounds,
+            mailbox_receiver,
+            journal_entries_per_section: cfg.journal_entries_per_section,
+            journal_replay_concurrency: cfg.journal_replay_concurrency,
+            journal_naming_fn: cfg.journal_naming_fn,
+            journals: HashMap::new(),
+            tip_man: TipManager::default(),
+            ack_man: AckManager::default(),
+            epoch: 0,
+        };
+
+        (result, mailbox)
+    }
+
+    /// Runs the actor until the runtime is stopped.
+    ///
+    /// The actor will handle:
+    /// - Timeouts
+    ///   - Rebroadcasting Links
+    ///   - Pruning old evidence
+    /// - Mailbox messages from the application:
+    ///   - Broadcast requests
+    ///   - Ack requests
+    /// - Messages from the network:
+    ///   - Links
+    ///   - Acks
+    pub async fn run(
+        mut self,
+        chunk_network: (impl Sender, impl Receiver),
+        ack_network: (impl Sender, impl Receiver),
+    ) {
+        let (mut link_sender, mut link_receiver) = chunk_network;
+        let (mut ack_sender, mut ack_receiver) = ack_network;
+        let mut shutdown = self.runtime.stopped();
+
+        loop {
+            // Enter the epoch
+            self.refresh_epoch();
```

Comment:
If you observe a threshold signature of an epoch greater than yours, you probably also want to move forward (not sure if you do that but thought of this in passing).

Reply:
That makes sense, however I think we should just wait for the application or coordinator to tell us to enter the next epoch. This means that it is ready to serve information like signers/sequencers/shares. If we just go ahead and enter the next epoch then that information will likely not be ready anyway, and there are more moving pieces to have to think about IMO